### PR TITLE
Feature/81 historical maps

### LINKED ISSUE #81: Add historical maps
- [ ] school page map
- [ ] search page map

### PR DESCRIPTION
**Pull request by lraivio**
This PR adds historical maps to school and search page maps.
Most of the previous map functionality had to be rewritten because I could not get the historical layers and proj4 to work with react-leaflet.
So, sorry for the big PR.

Known issues.
- There is a critical dependency warning when bundling the scripts. I guess this is because proj4-leaflet includes the proj4 package internally from a dist, not src folder. So the proj4 package is already prebuilt. I don't know if there is anything we can do about this. Or if there is a good way to not show that error in terminal.
- Zooming while changing historical layers on search map is not always optimal.
- There is no link to school pages from the search map anymore.
- All schools are shown on all layers. So they are not yet filtered based on the historical map years.

Closes #81.

### CODE REVIEW

**Comment by YoYuUm** [comment]
A rebase is needed.

File: app/styles/maps.less [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  .leaflet-popup-content-wrapper {
+    border: 2px solid @border-main-color;
+    color: @base-color;
+  }
+  .leaflet-popup-content {
+    font-size: 14px;
+    margin: 6px 20px;
+  }
+  .leaflet-popup-tip {
+    border: 2px solid @border-main-color;
```
Comment (YoYuUm) [inline app/styles/maps.less]: I would add a background-color: @base-color so there is not a line between the content and the tip.

**Comment by YoYuUm** [comment]
What should happen with schools out of the boundaries of historical maps? like http://localhost:3000/koulut/236 for example.

**Comment by YoYuUm** [comment]
Map zoom doesn't work sometimes when the address changes. Eg: http://localhost:3000/koulut/248

**Comment by lraivio** [comment]
There's not much we can do about school going out of boundaries. Hopefully we'll get more extensive historical maps in the future.

**Comment by lraivio** [comment]
The zoom problem is a strange one. I'll have to investigate that further to see what's going on :/

File: app/styles/maps.less (hunk 1 above)
Comment (lraivio) [inline app/styles/maps.less]: That's a good idea! I was only thinking about how to make the insides white but that's a good solution.

File: app/core/mapUtils.js [hunk 2]
```diff
@@ -1,6 +1,58 @@
...
+L.Map.prototype.setCrs = function(newCrs) {
+  this.options.crs = newCrs;
+};
+const leafletImageFolder = path.resolve(__dirname, '../../node_modules/leaflet/dist/images/');
+L.Icon.Default.imagePath = leafletImageFolder;
+
+const crs = _makeCRS();
+const layers = _makeLayers();
+
+function _getGeoerverUrl(layerName, layerFmt) {
```
Comment (YoYuUm) [inline app/core/mapUtils.js]: Typo

File: app/core/mapUtils.js [hunk 3]
```diff
@@ -1,6 +1,58 @@
...
+  /*eslint-enable */
+}
+
+function _makeCRS() {
+  const crsName = 'EPSG:3879';
+  /*eslint-disable */
+  const projDef = '+proj=tmerc +lat_0=0 +lon_0=25 +k=1 +x_0=25500000 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs';
+  /*eslint-enable */
+  const bounds = [25440000, 6630000, 25571072, 6761072];
+  const resolutions = [256, 128, 64, 32, 16, 8, 4, 2, 1, 0.5, 0.25, 0.125, 0.0625];
```
Comment (YoYuUm) [inline app/core/mapUtils.js]: Maybe bounds and resolutions could be app constants?

Comment (YoYuUm) [inline app/core/mapUtils.js]: Maybe it's not needed, because this file is mostly full of that kind of variables.

**Comment by YoYuUm** [comment]
Why there is no link to school pages from the search map ?

File: app/components/SearchMap.js [hunk 4]
```diff
@@ -1,96 +1,82 @@
...
+      .setView(HELSINKI_COORDINATES, MAP_ZOOM);
+    L.control.layers(this.layers).addTo(this.map);
+    this.markerGroup.addTo(this.map);
+
+    const self = this;
+    this.map.on('baselayerchange', function() {
+      const centerPoint = self.map.getCenter();
+      const zoom = MAP_ZOOM;
+      self.map.setView(centerPoint, zoom);
+    });
```
Comment (YoYuUm) [inline app/components/SearchMap.js]: You could have pass context using "( ) => " notation. That way you can use this inside the anonymous function and you don't need to declare self.

**Comment by YoYuUm** [comment]
Great job. I you have managed to get a good structure of the code.

**Comment by lraivio** [comment]
Currently there is no links to school pages because Link can't be used inside leaflet popups and a normal "a"-elements will cause a full page refresh. So I didn't implement the linking before coming up with a better solution.

**Comment by YoYuUm** [comment]
Ok, I will create a ticket for it then :)

File: app/components/SearchMap.js (hunk 4 above)
Comment (lraivio) [inline app/components/SearchMap.js]: Good idea!

**Comment by lraivio** [comment]
Thanks for the ticket :)

**Comment by lraivio** [comment]
I made the requested changes and rebased master. 
I also added an issue #95 for the bug you found about map zoom freezing on address change.

**Comment by lraivio** [comment]
I also realized that the search timeline should not be visible while on map view but maybe I can change that when I implement the map year selectors.

**Comment by YoYuUm** [comment]
:+1:

**Pull request merged**